lib/rbac/filterer.rb

```diff
@@ -424,25 +422,45 @@ def select_from_order_columns(columns)
       end
     end
 
-    def include_references(scope, klass, references, exp_includes)
-      scope.references(klass.includes_to_references(references)).references(klass.includes_to_references(exp_includes))
+    def include_references(scope, klass, includes, references, exp_includes)
+      if scope.respond_to?(:eager_load)
+        scope.eager_load(klass.prune_references(references)).eager_load(exp_includes || []).preload(includes)
+      else
+        # TODO: add eager_load/preload support to QueryRelation/ActsAsArModel
+        # Then drop this if block and includes_to_references
+        scope.references(klass.includes_to_references(references)).references(klass.includes_to_references(exp_includes))
+            .includes(includes).includes(exp_includes)
+      end
     end
 
     # @param includes [Array, Hash]
     def add_joins(klass, scope, includes)
       return scope unless includes
 
+      # NOTE: We should be pre-pruning polymorphic out of here. So they shouldn't come in
+      #       either way, want a distinct for collections (to safeguard has_many values).
+      # NOTE: This does not recurse, so not a complete implementation for pruning/distinct
       includes = Array(includes) unless includes.kind_of?(Enumerable)
       includes.each do |association, value|
         reflection = klass.reflect_on_association(association)
         if reflection && !reflection.polymorphic?
           scope = value ? scope.left_outer_joins(association => value) : scope.left_outer_joins(association)
-          scope = scope.distinct if reflection.try(:collection?)
         end
       end
+      scope = scope.distinct if !scope.distinct_value && is_collection?(klass, includes)
       scope
     end
 
+    def is_collection?(klass, includes)
+      return false unless includes.present?
+
+      Array(includes).each do |association, value|
+        reflection = klass.reflect_on_association(association)
+        return true if reflection&.collection? || !reflection.polymorphic? && is_collection?(reflection.klass, value)
+      end
+      false
```

Comment:
Simpler?

```suggestion
      Array(includes).any? do |association, value|
        reflection = klass.reflect_on_association(association)
        reflection&.collection? || !reflection.polymorphic? && is_collection?(reflection.klass, value)
      end
```

Reply:
this is a `break` from the iterator.

I will see if we can convert to a `detect?`

Reply:
thanks. this turned out well